src/symbol/projection.js

```diff
@@ -0,0 +1,249 @@
+        }
+
+        const cameraToAnchorDistance = anchorPos[3];
+        const perspectiveRatio = 1 + 0.5 * ((cameraToAnchorDistance / painter.transform.cameraToCenterDistance) - 1);
+
+        const fontSize = symbolSize.evaluateSizeForFeature(sizeData, partiallyEvaluatedSize, symbol);
+        const pitchScaledFontSize = pitchWithMap ?
+            fontSize * perspectiveRatio :
+            fontSize / perspectiveRatio;
+        const fontScale = pitchScaledFontSize / 24;
+
+        const glyphsForward = [];
+        const glyphsBackward = [];
+
+        const end = symbol.glyphStartIndex + symbol.numGlyphs;
+        for (let glyphIndex = symbol.glyphStartIndex; glyphIndex < end; glyphIndex++) {
+            const glyph = bucket.glyphOffsetArray.get(glyphIndex);
+            if (glyph.offsetX > 0) {
+                glyphsForward.push(glyph);
+            } else {
+                glyphsBackward.push(glyph);
+            }
+        }
+        processDirection(glyphsForward, 1, flip, symbol, lineVertexArray, dynamicLayoutVertexArray, labelPlaneMatrix, fontScale);
+        processDirection(glyphsBackward, -1, flip, symbol, lineVertexArray, dynamicLayoutVertexArray, labelPlaneMatrix, fontScale);
+    }
+
+    if (isText) {
+        bucket.buffers.glyph.dynamicLayoutVertexBuffer.updateData(dynamicLayoutVertexArray.serialize());
+    } else {
+        bucket.buffers.icon.dynamicLayoutVertexBuffer.updateData(dynamicLayoutVertexArray.serialize());
+    }
+}
+
+function processDirection(glyphs, dir, flip, symbol, lineVertexArray, dynamicLayoutVertexArray, labelPlaneMatrix, fontScale) {
+    assert(symbol.lineLength > 1);
+
+    let prev = project(new Point(symbol.anchorX, symbol.anchorY), labelPlaneMatrix);
+    let next = prev;
+    let vertexIndex = 0;
+    let previousDistance = 0;
+    let segmentDistance = 0;
+    let segmentAngle = 0;
+
+    let numVertices, vertexStartIndex;
+    let angle = 0;
+
+    if (flip) {
+        // The label needs to be flipped to keep text upright.
+        // Iterate in the reverse direction.
+        dir *= -1;
+        angle = Math.PI;
+    }
+
+    if (dir === 1) {
+        numVertices = symbol.lineLength - symbol.segment - 1;
+        vertexStartIndex = symbol.lineStartIndex + symbol.segment + 1;
+    } else {
+        numVertices = symbol.segment + 1;
+        vertexStartIndex = symbol.lineStartIndex + symbol.segment;
+        angle += Math.PI;
+    }
+
+    // For each glyph, find the point `offsetX` distance from the anchor.
+    for (const glyph of glyphs) {
+        const offsetX = Math.abs(glyph.offsetX) * fontScale;
+
+        // If the current segment doesn't have enough remaining space, iterate forward along the line.
+        // Since all the glyphs are sorted by their distance from the anchor you never need to iterate backwards.
+        // This way line vertices are projected at most once.
+        while (offsetX >= segmentDistance + previousDistance && Math.abs(vertexIndex) < numVertices) {
```

Comment:
This while loop exits if we reach the end of the line (`vertexIndex >= numVertices`), but we keep adding glyphs onto whatever the last segment was. This leads to some funky behavior:

- Ignoring collisions (because collision boxes only go as long as the line)
- Road labels extending past the end of the road
- Road labels wrapping back on themselves (see Ocean Park Blvd in the screen capture)... although this one might be better addressed by a `max-curvature` property.

![turnabout is fair play](https://user-images.githubusercontent.com/375121/26846147-aee13f40-4aad-11e7-8e7c-b544689629e9.gif)

We should hide the label instead if we hit this limit.